### ISSUE TITLE
Consider inverting model and dispatch arguments for the view

### ISSUE THREAD

**Issue by vbfox** [elmish/elmish#130]
### Description

The current order of model and dispatch for the view function mean that the function can't capture dispatch and generate functions once.

I don't know for other Elmish implementations but for React it's a perf problem.

The code must generate action functions on each call of the view and it break all optimizations that something like PureComponent provide and force all DOM elements that have event handler to need to be changed, poluting the diff algorithm.

### Repro code

```fsharp
type [<Pojo>] BookProps = { key: string; id: string; book: string; removeBook: string-> unit }

type BookComponent(initialProps) =
    inherit PureStatelessComponent<BookProps>(initialProps)
	member this.removeBook() = this.props.removeBook this.props.id
    override this.render() =
        tr [] [
            td [] [ str this.props.book ]
            td [] [ R.a [ OnClick this.removeBook] [ str "Remove" ] ]
        ]

let view (model:Model) (dispatch: Msg -> unit) =
    let removeBook id = dispatch (RemoveBook id)
    table [] [
        model.WishList.Books
            |> List.map(fun book -> BookComponent { key = book.id; id = book.id; book = book.title; removeBook = removeBook })
            |> ofList
    ]
```

While we follow all of React best practices, the PureComponent is useless there as `removeBook` changes each time the view is called 😭 

### Expected and actual results

I expect the library to lend me in the pit of success and allow me to do get the best perf with ease 😉 

With the arguments inverted the view would look like that (With `Program.Run` changed to cache the result of passing just dispatch) :

```fsharp
let view (dispatch: Msg -> unit) =
    let removeBook id = dispatch (RemoveBook id)
    fun (model:Model) ->
        table [] [
            model.WishList.Books
                |> List.map(fun book -> BookComponent { key = book.id; id = book.id; book = book.title; removeBook = removeBook })
                |> ofList
        ]
```

React will always see the same props and optimize the hell out of it (And i'll ideally introduce a `BookListComponent` that is also pure so that `render()` is never called if books didn't change making it ultra cheap when books don't change.

### The dirty Hack (TM)

Some hack like that could work to make it resonably fast in the current state (No tested) but that's ugly:

```fsharp
type HackData<'msg, 'model, 'view> = {
    dispatch: 'msg -> unit
    captured: 'model -> 'view
}

let mkHackView (from: ('msg -> unit) -> 'model -> 'view): 'model -> ('msg -> unit) -> 'view =
    let mutable current: HackData<'msg, 'model, 'view> option = None
    fun model dispatch ->
        if (current.IsNone || (not (obj.ReferenceEquals(current.Value.dispatch, dispatch)))) then
            current <- Some { dispatch = dispatch; captured = from dispatch }

        current.Value.captured model
```

### End remark

I know that it's a big change but the possibilities it open are very interesting and I don't see any way to generate good React code without it.

(But maybe i'm compleltely wrong, please tell me if there is a magic way ✨ )

**Comment by MangelMaxime** [comment]
Would this still works with extra arguments ?

```fs
let view (dispatch: Msg -> unit) =
    let removeBook id = dispatch (RemoveBook id)
    fun (user : User) (model:Model) ->
        table [] [
            model.WishList.Books
                |> List.map(fun book -> BookComponent { key = book.id; id = book.id; book = book.title; removeBook = removeBook })
                |> ofList
        ]
```

Note: `fun (user : User) (model:Model) ->`

From my experience, view is rarely just `Msg -> Model -> React.HTMLElement`

Also, we will need to make users understand that:

```fs
let view (dispatch: Msg -> unit) =
    let removeBook id = dispatch (RemoveBook id)
    fun (model:Model) ->
        table [] [
            model.WishList.Books
                |> List.map(fun book -> BookComponent { key = book.id; id = book.id; book = book.title; removeBook = removeBook })
                |> ofList
        ]
```

is different from:

```fs
let view (dispatch: Msg -> unit) (model : Model) =
    let removeBook id = dispatch (RemoveBook id)

    table [] [
        model.WishList.Books
            |> List.map(fun book -> BookComponent { key = book.id; id = book.id; book = book.title; removeBook = removeBook })
            |> ofList
    ]
```

It's some kind of "advanced" knowledge.

**Comment by alfonsogarciacaro** [comment]
Thanks a lot for looking into these issues and help improve Elmish performance @vbfox!

I think the change makes sense (though it's a pity that breaks the API). When writing a custom `view` function I often change the order of the arguments because you usually want to have the model at the end (e.g. `myData |> List.map (view dispatch)`). However I'm concerned that this would still depend on the developer's discipline, I frequently pass the dispatch function directly to children component like `(Msg >> dispatch)` so I guess the optimization won't work in that case. Also, if this depends on `Program.run` it will only affect the top level of your app, right?

Maybe providing an alternative API more explicit towards `dispatch` catching?

```fsharp
let optimizeView (view: 'Model->('Msg2->unit)->'View)
                 (transformDispatch: ('Msg->unit)->'Msg2->unit) =
    let mutable dispatchCache = None
    fun model dispatch ->
        let dispatch2 =
            match dispatchCache with
            | Some d -> d
            | None ->
                let d = transformDispatch dispatch
                dispatchCache <- Some d
                d
        view model dispatch2
```

Something like could also prevent problems with HMR, by not caching dispatch in DEBUG mode for example.

**Comment by et1975** [comment]
Good observation, callback diffing totally skipped my mind when I designed this! 
However, if you know the model hasn't changed you can skip React diffing entirely with `lazyView`. I'm open to a breaking change in general, but simply reversing the arguments would not quite do what you expect, because the caller would also need to curry the `dispatch` in early.

**Comment by alfonsogarciacaro** [comment]
I was thinking also of `lazyView`, though I guess it cannot be generalized for all situations as here you'd have to use `lazyViewWith` with a custom equality function that ignores the `removeBook` field.

@vbfox Do you know if there's any advantage of using `PureStatelessComponent` instead of a component implementing `shouldComponentUpdate` (what `lazyView` does)?

**Comment by et1975** [comment]
@alfonsogarciacaro `lazyView` does not diff the views, it diffs the models, so it would not be affected by the different callback.

**Comment by alfonsogarciacaro** [comment]
@et1975 Ah, true! I was thinking that in this case @vbfox was passing `removeBook` through the props, but if you use `lazyView` you probably wouldn't do that, rather:

```fsharp
type BookModel = { key: string; id: string; book: string }
type BookMsg = RemoveBook of string

let bookView dispatch (model: BookModel) =
  tr [] [
      td [] [ str model.book ]
      td [] [ a [ OnClick (fun _ -> RemoveBook model.id |> dispatch)] [ str "Remove" ] ]
  ]

let view model dispatch =
    table [] [model.Books |> List.map (bookView dispatch) |> ofList]
    // or:
    // table [] (model.Books |> List.map (bookView dispatch))
```

The only problem is `lazyView` doesn't contain `key` so it can be problematic to use it in lists. It may be a good idea to [add it](https://github.com/fable-elmish/react/blob/71fa3f60e61b1cb941c8d988d32871725f1eeeca/src/common.fs#L35)

**Comment by et1975** [comment]
@alfonsogarciacaro 
> The only problem is `lazyView` doesn't contain key so it can be problematic to use it in lists. It may be a good idea to add it.
I vaguely recall considering it, but I can't remember why I didn't implement it. Maybe because there was no good way to avoid specifying the key explicitly? Can we move `key` prop from the (embedded) view into the (wrapping) lazyView?

**Comment by alfonsogarciacaro** [comment]
Using dynamic programming, yes, though it'd be less clean. The other possibility off the top of my head is using options. The option is erased for string and for Fable 2.0 I was thinking to remove fields initialized to `None` in records.

**Comment by vbfox** [comment]
> Would this still works with extra arguments ?

Yes it shouldn't change a lot in this case, change the case where you want to call another view from your view but once the pattern is there it can be applied everywhere.

```fsharp
let view (dispatch: Msg -> unit) =
    let otherView = otherView dispatch

    fun  (model : Model) -> 
    table [] [
        otherView model.subPart
    ]
```

> I'm open to a breaking change in general, but simply reversing the arguments would not quite do what you expect, because the caller would also need to curry the dispatch in early.

Yes I already looked at what it would involve and it's not much change in the core I think.

> It's some kind of "advanced" knowledge.

> However I'm concerned that this would still depend on the developer's discipline, I frequently pass the dispatch function directly to children

Yes, the fact that you should never use lambdas in React `render()` is too (Even if it's well documented on blogs by this point)

People can be guided toward that design on a second time after learning the elmish principles. As all optimizations they should be considered in context, I suggest the change to make it easy to use the optimization when needed.

> Maybe providing an alternative API more explicit towards dispatch catching?

That's pretty much my hack. It's totally doable but except for the pain of the act of changing it, the library could totally do it.

> Do you know if there's any advantage of using PureStatelessComponent instead of a component implementing shouldComponentUpdate (what lazyView does)?

`lazyView` can nearly work perf-wise if used at each level but it make UI harder to debug as the whole react debugger is filled with LazyView2 and the props aren't nicely displayed as everything is in the `model` member.

Two other notes :
* While lazyView can solve some of the problems of calling render() it doesn't solve the problem currently created if render IS called: Every event handler would change and that would bubble to the DOM diff and produce useless DOM changes... And that's not an impossible case for example if you receive objects from WebSocket they are different each time so no cheap ReferenceEquals comparison (And even the fsharp record equality might not be an option in fable 2.0)
* Maybe I'm biased but I'm thinking of developers with React experience that would already know such patterns and move from js/ts/flow to fable and would expect for them to apply.

**Comment by OlegZee** [comment]
According to "Domain Modeling Made Functional" book by Scott Wlaschin the dependencies should go before function arguments. To me the dispatch is a pure dependency so dispatch should go first. Just another reason for making this change.

**Comment by et1975** [comment]
>Yes I already looked at what it would involve and it's not much change in the core I think.

The core needs not change at all, I think, as `setState` is already offloaded to Elmish.React. The changes there are minimal, but that's not all that interesting, as the root view is always wrapped in `lazyView` and you already get the optimization that skips the render entirely. Like @alfonsogarciacaro said, in order to benefit from this the currying would have to happen at nested view levels, which you could accomplish today with some creative memoization. 

Regardless, I think it's a good change 👍 , how is everyone else feeling?
Longer term, what do you folks think of removing the `dispatch` arg from the view entirely (a la Elm)?

As long as we are making breaking changes, how does everybody feel about reversing the order of msg and model in `update`? 😱 

This could make update look like this:
```fsharp
let update model = function
     | Msg1 m -> ...//
     | Msg2 m -> ...//
```

*hides* 👀

**Comment by vbfox** [comment]
> The core needs not change at all, I think, as setState is already offloaded to Elmish.React. 

Why not change the core ? wouldn't it be confusing to have the args in one order in the core an in another in the react part ?
Also there react part would need to implement an equivalent to my "hack", and any other `Elmish.XXX` would have to re-do that where in the core it could be implemented easily and spread with small changes to all sub projects.

> Longer term, what do you folks think of removing the dispatch arg from the view entirely (a la Elm)?

I'm not sure it's possible to do that and keep being able to use external react components easily.

----

Well my personal breaking change would be to stop having commands be functions and instead have them be a Discriminated Unions with an additional `'cmd -> unit` command runner in program. That would allow unit testing of Elmish programs but i'm not sure it's worth it.

**Comment by vbfox** [comment]
> As long as we are making breaking changes, how does everybody feel about reversing the order of msg and model in update?

I like that it make the impl cleaner

**Comment by et1975** [comment]
@vbfox
>That would allow unit testing of Elmish programs but i'm not sure it's worth it.

Does #123 discussion not address your concerns? I believe Elmish programs are supremely testable, so I'd be curious to know how is that not the case for you (please comment in #123).

**Comment by et1975** [comment]
>  in the core it could be implemented easily and spread with small changes to all sub projects.

Fair enough. Would you be sending the PRs?

**Comment by vbfox** [comment]
Yes i'll be sending the PR

> Does #123 discussion not address your concerns?

No but i'll comment there

**Comment by MangelMaxime** [comment]
I am not sure of the choice to remove the `cmd` I kind of like them because, they force us to handle success and fail cases.

How would we dispatch a message from our view without the `dispatch` function ? For example, in some of my application, I am using `Ref` attribute to trigger specific action and then use `dispatch` to notify the model that I did this action.

**Comment by et1975** [comment]
> How would we dispatch a message from our view without the dispatch function ?

I'm not sure how exactly Elm does its majik, but the callbacks introduced by your view would return your message instead of `unit`. Presumably, we'd introduce custom React DSL that redefines `Props` walk the props list to convert Elmish.React props into React props (injecting the dispatch calls whenever a message-returning prop is encountered). Approached this way it would be a nightmare to integrate with React libs, no doubt, so I'm not seriously considering it. But perhaps there's a nicer way?

**Comment by MangelMaxime** [comment]
Was thinking the same as you that's why I asked ^^.

And for now, I don't have another idea. The "problem" that Elm don't have is the integration with an existing rendering lib (react)

**Comment by alfonsogarciacaro** [comment]
@vbfox 

> lazyView can nearly work perf-wise if used at each level but it make UI harder to debug as the whole react debugger is filled with LazyView2

Ah true, I don't use the React tools often but I remember I saw that once and wasn't very nice. Apparently [there's a way to change the display name of the component in React tools](https://reactjs.org/docs/higher-order-components.html#convention-wrap-the-display-name-for-easy-debugging), maybe we could use that?

> Yes, the fact that you should never use lambdas in React render() is too (Even if it's well documented on blogs by this point)

I didn't know you shouldn't use lambdas in `render()` 😅 It could be possible to introduce an optimization in Fable so in certain contexts lambdas are extracted so they're no generated every time (actually someone suggested that once). That would be neat and could be a good selling point for Fable.

```fsharp
let view x =
   foo (fun y -> y + x)
```

```js
function $auto123(x) {
   return function (y) {
      return y + x; 
   }
}

function view(x) {
  return foo($auto123(x));
}
```

**Comment by alfonsogarciacaro** [comment]
BTW, in Fable 2.0 records and unions will still have structural equality, what will likely be not possible any more is custom equality by overriding `Equals` (it will still be ok for classes).

**Comment by vbfox** [comment]
> I didn't know you shouldn't use lambdas in `render()`

It break PureComponent for the same reason as here: reference equality for delegates.

It's not in most react tutorials as it's less important for "Raw React" because the root component is nearly never re-mounted so updates are always local (done by calling setState)
But for Redux/Elm patterns where each store change mean a full render from the root and the optimization come from stopping the `render()` cascade via `shouldComponentUpdate` it's the only thing stopping each change from re-creating ALL event handlers on the page.

**Comment by vbfox** [comment]
Also a `lazyView` with `displayName` would be very sweet, that's true.

**Comment by alfonsogarciacaro** [comment]
Scratch my comment above, I just realized that would still create a new closure every time so it's no use. Sorry, I know, I should got to sleep 😴 


const $auto123(x) =
   return function (y) {
      return y + x; 
   }
}

function view(x) {
  return foo($auto123(x));
}

**Comment by et1975** [comment]
2.0 Core is out, keeping this open until everything in Elmish stack has been migrated.

**Comment by et1975** [comment]
Should I expect Elmish.React PR as well?

**Comment by et1975** [comment]
Kinda regret switching the target to netstandard2.0, I'll go back to 1.6 for Fable build... unless @alfonsogarciacaro plans to move everything forward anyway?

**Comment by alfonsogarciacaro** [comment]
There are no plans to move to netstandard2.0, but as FSharp.Core has moved now it may make sense to align with it.

**Comment by et1975** [comment]
@vbfox 

I run [some tests](/et1975/react-fable-elm-performance-comparison) and I have a few observations that make me consider scraping the 2.0 release:
- Even in the mickey-mouse TodoMVC it's a lot harder than I imagined it would be to optimize around curried dispatch - aside from adding tons of code, most event-handlers need something from the model embedded in the closure/context.
- With the caveat that most of the handlers remain unoptmized the performance gains so far are minimal. Breaking it down between `lazy` and RAF, RAF accounts for about 20% improvement, so switching the default to "sans RAF" might be a bad move.

Here are some numbers from my OSX/Chrome MBP (the pink is optimized, blue - unoptimized):

![image](https://user-images.githubusercontent.com/623703/37931378-6d40d30a-3113-11e8-9654-062461f1e3d9.png)

>Offtopic: I've run this benchmark on many different browsers and versions and it's always amazing how different the results are from one to another. Here the current Chrome's doing something seriously weird to make optimized Elm version slower!

I hope you can make some improvements to the sample to present the changes in a better light. As it stands now, I feel like some sort of memoization might be a better option.

**Comment by MangelMaxime** [comment]
At some point, I tried to port my production App to Elmish 2.0 beta and to be fair the code is really verbose and it's really easy to not optimize the code.

I do agree about:

> most event-handlers need something from the model embedded in the closure/context.

I can understand the benefit of being 100% compliant with React, but with current proposition I am not sure of the benefit.

We can also, consider that force the user to keep using Elmish the same way we do now even with Elmish 2.0 proposition. But then, we could have a split on how users code in Elmish (for blogs, samples, etc.) and can be harmful.

Last thing, my current application at work have some slow time that because didn't included any optimization in it. And when comparing Elmish 1.0 and Elmish 2.0 performances then are kind of the same, I don't have any exact number and will not be able to provide then because I reverted my changes.

I think some example about `memoization` could be better in the long term.
